File: MATLAB/PMC.m
function [C,cmax] = PMC(adjacency_matrix,core_numbers)

A = sparse(adjacency_matrix);
c_n = core_numbers;
 n = size(A, 1); % Number of vertices
 
    
    % Initialize variables
    C = []; 
    cmax = 0;
    
    % Sort vertices by core number in descending order
    [~, sorted_indices] = sort(c_n, 'descend');
    sorted_vertices = sorted_indices';
    
    % Loop over sorted vertices
    for i = 1:n
        vi = sorted_vertices(i);
        
        if c_n(vi) >= cmax
            % Neighbors of vi with core numbers >= cmax
            S = find(A(vi,:) & (c_n >= cmax)');
            
            % Sort neighbors in descending order by core number
            [~, sorted_neighbors] = sort(c_n(S), 'descend');
            sorted_neighbors = S(sorted_neighbors);
            
            % Initialize C'
            C_prime = [vi];
            
            % Loop over sorted neighbors
            for j = 1:length(sorted_neighbors)
                sj = sorted_neighbors(j);
                
                % Check if sj is connected to all vertices in C'
                
                if all(A(C_prime, sj))
                    C_prime = [C_prime, sj]; % Add sj to C'
                end
                
                % If C' forms a larger clique, update C and cmax
                if length(C_prime) > cmax
                    C = C_prime;
                    cmax = length(C_prime)-1;
                end
            end
        end
    end
  
    


end

File: cpp/Src/pmc.cpp
#include "pmc.hpp"


bool isClique(const std::vector<std::vector<int>>& adjacencyMatrix, const std::vector<int>& C_prime, int sj) {
    return std::all_of(C_prime.begin(), C_prime.end(), [&](int vertex) {
        return adjacencyMatrix[vertex][sj];
    });
}

std::vector<int> PMC(const std::vector<std::vector<int>>& adjacency_matrix, const std::vector<int>& c_n){

int n = adjacency_matrix.size();
std::vector<int> S{};
std::vector<int> C_prime;
std::vector<int> C{};

std::vector<int> sorted_vertices;

 for (size_t i{}; i < c_n.size(); ++i) {
        sorted_vertices.push_back(i);
    }
sort(sorted_vertices.begin(), sorted_vertices.end(), [&](int a, int b){
     return c_n[a] >= c_n[b];
    });
     

int cmax{};

    // Loop over sorted vertices
    for(size_t i{}; i < sorted_vertices.size(); ++i){
       
        int v = sorted_vertices[i];

        S = {};

        

        if (c_n[v] >= cmax){
             for (size_t j{}; j < adjacency_matrix.size(); ++j) {
                if (adjacency_matrix[v][j] == 1 && c_n[j] >= cmax) {
                    S.push_back(j);
                }}
              
                 sort(S.begin(), S.end(), [&](int a, int b){
                return c_n[a] >= c_n[b];
            });
           

            C_prime = {v};
            
            
            // Loop through vertices in sorted S
            for (size_t k{}; k < S.size(); ++k) {
            
        
                
                

                    if (isClique(adjacency_matrix, C_prime, S[k])) {
                          C_prime.push_back(S[k]);
                        
                    }

                  
           if (C_prime.size() > cmax) {
                C = C_prime;
                cmax = C_prime.size() - 1;
            }
        }
    }
    
        
    }
        
        
         return C; 
    }
